scopesim/utils.py

```diff
@@ -1064,3 +1069,394 @@ def return_latest_github_actions_jobs_status(owner_name="AstarVienna", repo_name
+    pix_in_reference_frame
+        Transforms pixel positions from one reference frame to another.
+    """
+
+    def __init__(self, n: int, *, x0: float = 0,  i0: Optional[int] = None, dx: float = 1.0):
+        """
+        Parameters
+        ----------
+        n : int
+            The dimension of the axis.
+        x0 : float
+            The central corodinate - default `0``
+        i0 : Optional[int]
+            The central pixel coordinate - default `n//2`.
+        dx : float
+            The pixel spacing - default `1.0`.
+        """
+        self.n = n
+        self.x0 = x0
+        self.i0 = i0 if i0 else n // 2
+        self.dx = dx
+
+    def pos_to_pix(self, pos: Union[float,  npt.NDArray]) -> npt.NDArray:
+        """Transforms real reference frame position to pixel positions.
+
+        Parameters
+        ----------
+        pos :
+            A single or an array of positions in the reference frame.
+
+        Returns
+        -------
+        npt.NDArray
+            The pixel position corresponding to the provided reference frame position.
+        """
+        pos = np.array(pos, ndmin=1)
+        return (pos - self.x0) / self.dx + self.i0
+
+    def pix_to_pos(self, pix: Union[float, npt.NDArray]) -> npt.NDArray:
+        """Transforms pixel positions to real reference frame positions.
+
+        Parameters
+        ----------
+        pix : Union[float, npt.NDArray]
+            A single or an array of pixel positions.
+
+        Returns
+        -------
+        npt.NDArray
+            The position in the reference frame corresponding to the provided pixel positions.
+        """
+        pix = np.array(pix, ndmin=1)
+        return self.x0 + (pix - self.i0) * self.dx
+
+    def pix_in_reference_frame(self,
+                               other: GridAxis,
+                               pix: Optional[Union[float,  npt.NDArray]] = None) -> npt.NDArray:
+        """Transforms pixel positions from one reference frame to another.
+
+        Parameters
+        ----------
+        other : GridAxis
+            The coordinate axis onto which the pixel coordinates are projected.
+        pix : Optional[Union[float,  npt.NDArray]]
+            The pixel positions to be evaluated. If not provided, all pixel positions are evaluated.
+
+        Returns
+        -------
+        npt.NDArray
+            The pixel coordinates projected in the other reference frame.
+        """
+        pix = np.array(pix, ndmin=1) if pix else np.arange(self.n)
+        return (self.x0 + (pix - self.i0) * self.dx - other.x0)/other.dx + other.i0
+
+    @property
+    def coors(self) -> npt.NDArray:
+        """Axis coordinates"""
+        return self.pix_to_pos(np.arange(self.n))
+
+    @property
+    def pix(self) -> npt.NDArray:
+        """Axis pixel coordinates"""
+        return np.arange(self.n) - self.i0
+
+    @property
+    def shape(self) -> Tuple[float]:
+        """Axis shape"""
+        return self.n,
+
+
+class GridCoordinates:
```

Comment:
I am not very familiar with WCS and more so with the `astropy` implementation, but I'll take a look when I have the chance. I also believe that my implementation works for now, but if WCS is well suited for this then it would be interesting do use it to avoid more custom code nad increase the probability of errors.